### PR TITLE
[SW-602] Introduce light REST API call to get frame info to fill Rapids frame cache

### PR DESCRIPTION
**Pull request by mmalohlava**
TODO:
  - [ ] discuss compatibility of REST API - the PR changes default from 0 to -1
  - [ ] need option to enable this feature on demand
  - [ ] we need to discuss generic strategy

### CODE REVIEW

File: h2o-core/src/main/java/water/api/schemas3/FrameV3.java [hunk 1]
```diff
@@ -207,31 +207,39 @@ public void clearBinsField() {
...
 
     row_count    = (int) Math.min(row_count, row_offset + f.numRows());
     column_count = Math.min(column_count, column_offset + f.numCols());
 
     this.frame_id = new FrameKeyV3(f._key);
-    this.checksum = f.checksum();
-    this.byte_size = f.byteSize();
+    if (expensive) {
+      this.checksum = f.checksum();
+      this.byte_size = f.byteSize();
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/api/schemas3/FrameV3.java]: goot point!

File: h2o-core/src/main/java/water/init/NetworkInit.java [hunk 2]
```diff
@@ -98,7 +98,6 @@ public static void initializeNetworkSockets( ) {
         break;
 
       } catch (Exception e) {
-        e.printStackTrace();
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/init/NetworkInit.java]: This seems not related to this change. Just to understand - the purpose is to clean up ? :)

**Comment by jakubhava** [comment]
Can confirm that this works with Sparkling Water & sparse data (just build sw against this pr and tested). It's already way faster. The delay of obtaining the head of data to show can be now configured on sw side by limiting number of columns to print.

File: h2o-core/src/main/java/water/init/NetworkInit.java (hunk 2 above)
Comment (mmalohlava) [reply h2o-core/src/main/java/water/init/NetworkInit.java]: yes. else we have annoying stack trace all the time in SW.

Comment (jakubhava) [reply h2o-core/src/main/java/water/init/NetworkInit.java]: thanks for explanation!

File: h2o-core/src/main/java/water/api/schemas3/FrameV3.java [hunk 3]
```diff
@@ -207,31 +207,39 @@ public void clearBinsField() {
...
+    return fillFromImpl(f, 1, (int)f.numRows(), 0, -1, false);
   }
 
-  public FrameV3 fillFromImpl(Frame f, long row_offset, int row_count, int column_offset, int column_count) {
-    if( row_count == 0 ) row_count = 100;                                 // 100 rows by default
-    if( column_count == 0 ) column_count = f.numCols() - column_offset; // full width by default
+  private FrameV3 fillFromImpl(Frame f, long row_offset, int row_count,
+                              int column_offset, int column_count,
+                              boolean expensive) {
+    if( row_count < 0 ) row_count = 100;                                 // 100 rows by default
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/schemas3/FrameV3.java]: This could be an issue where

**Comment by michalkurka** [comment]
@mmalohlava Can we make the changes in Python optional, that is: introduce an option that would let us switch between the old and new behavior (same as with the "fusions")?

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by mmalohlava** [comment]
@michalkurka yup

**Comment by mmalohlava** [comment]
The biggest problem is probably my change of defaults for fetch call:
  -  `0`rows was interpreted as `100`, i changed that to interpret `-1` as `100`, since the caller does not need to be interested to see always 100rows, if it needs to fetch chunk_summary for exampple.
  - the same for columns: before `0` -> `all columns`, now: `-1` -> `all columns`

**Comment by mmalohlava** [comment]
@h2o-ops please test!

**Comment by michalkurka** [comment]
@mmalohlava What do you think about introducing "LightFrameV3" in the object hierarchy of FrameV3? This would help us get rid off the "expensive" flag and it would nicely specify what to expect on the client side.

**Comment by mmalohlava** [comment]
@h2o-ops please test!

File: h2o-core/src/main/java/water/api/schemas3/FrameV3.java (hunk 3 above)
Comment (mmalohlava) [reply h2o-core/src/main/java/water/api/schemas3/FrameV3.java]: yup, this is breaking REST API contract, but we can fix it on client side (send -1 instead of 0).

**Comment by mmalohlava** [comment]
@michalkurka initially i tried to avoid to make a new element in hierarchy (it is already convoluted), but worth trying now. CC: @jakubhava

File: h2o-py/tests/testdir_jira/pyunit_sw_602.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+data = [numpy.arange(0, 50000).tolist() for x in numpy.arange(0, 99).tolist()]
+fr = h2o.H2OFrame(data)
+full = H2OFrame.get_frame(fr.frame_id)
+light = H2OFrame.get_frame(fr.frame_id, light=True)
+
+assert full._ex._cache._id == light._ex._cache._id
+assert full._ex._cache._nrows == light._ex._cache._nrows
+assert full._ex._cache._ncols == light._ex._cache._ncols
+assert full._ex._cache._names == light._ex._cache._names
+assert full._ex._cache._data == light._ex._cache._data
```
Comment (jakubhava) [inline h2o-py/tests/testdir_jira/pyunit_sw_602.py]: Python can compare dicts as we expect

File: h2o-py/tests/testdir_jira/pyunit_sw_602.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
```
Comment (michalkurka) [inline h2o-py/tests/testdir_jira/pyunit_sw_602.py]: The formatting of this file seems off. Is it only because of github?

File: h2o-core/src/main/java/water/api/schemas3/FrameV3.java (hunk 3 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/api/schemas3/FrameV3.java]: So do we need to change R client as well?

File: h2o-py/tests/testdir_jira/pyunit_sw_602.py (hunk 5 above)
Comment (jakubhava) [reply h2o-py/tests/testdir_jira/pyunit_sw_602.py]: good point, let me fix it

**Comment by jakubhava** [comment]
Closing this one in favour of https://github.com/h2oai/h2o-3/pull/1808 in rel-wheeler for now
